bip-xxxx.mediawiki

```diff
@@ -0,0 +1,344 @@
+<pre>
++----------+                        +--------+         +-----------------+
+| Receiver |                        | Sender |         | Bitcoin Network |
++----+-----+                        +---+----+         +-------+---------+
+     |       +-----------------+        |                      |
+     +-------+ BIP21 with ?pj= +------->+                      |
+     |       +-----------------+        |                      |
+     |                                  |                      |
+     |        +---------------+         |                      |
+     +<-------+ Original PSBT +---------+                      |
+     |        +---------------+         |                      |
+     |                                  |                      |
+     |       +------------------+       |                      |
+     |       | Payjoin Proposal |       |                      |
+     +-------+      PSBT        +------>+                      |
+     |       +------------------+       |                      |
+     |                                  |   +--------------+   |
+     |                                  |---+ Payjoin      |   |
+     |                                  |   | transaction  +-->+
+     |                                  |   +--------------+   |
+     +                                  +                      +
+</pre>
+The original PSBT is sent in the HTTP POST request body, base64 serialized, with <code>text/plain</code> in the <code>Content-Type</code> HTTP header and <code>Content-Length</code> set correctly.
+The payjoin proposal PSBT is sent in the HTTP response body, base64 serialized with HTTP code 200.
+
+To ensure compatibility with web-wallets and browser-based-tools, all responses (including errors) must contain the HTTP header <code>Access-Control-Allow-Origin: *</code>.
+
+The sender must ensure that the url refers to a scheme or protocol using authenticated encryption, for example TLS with certificate validation, or a .onion link to a hidden service whose public key identifier has already been communicated via a TLS connection. Senders SHOULD NOT accept a url representing an unencrypted or unauthenticated connection.
+
+===Receiver's well known errors===
+
+If for some reason the receiver is unable to create a payjoin proposal, it will reply with a HTTP code different than 200.
+The receiver is not constrained to specific set of errors, some are specified in this proposal.
+
+The errors have the following format:
+<pre>
+{
+    "errorCode": "leaking-data",
+    "message": "Key path information or GlobalXPubs should not be included in the original PSBT."
+}
+</pre>
+
+The well-known error codes are:
+{| class="wikitable"
+!Error code
+!Meaning
+|-
+|leaking-data
+|Key path information or GlobalXPubs should not be included in the original PSBT.
+|-
+|psbt-not-finalized
```

Comment:
What's the point of errors like this? Wouldn't it better for the receiver to just use `invalid-transaction` ? Any sender smart enough to handle an error like this, is also smart enough to not make the error in the first place.

Reply:
So you would remove `insane-psbt`, `psbt-not-finalized` `leaking-data` and `need-utxo-information` ?

Reply:
Yeah, I would remove them all. I have always regretted adding errors for cases that shouldn't happen. Like there's just no way for someone to write a program that handles them -- or at least I can't think of one. When ever designing APIs I now always have a free-form "counterparty error" style that I use, which I find very convenient and normally write my code such that when in dev/debug build, I provide a huge amount of details on the error and in production builds are much quieter with what I reveal. 

The other advantage of this style is it allows you to make sure comprehensively handle all the different error codes with a sane code path.

Reply:
It's not possible to enumerate all the ways the sender will screw up, so just having a single "you screwed up" error with a free form description is a lot more practical. And it's a code-path that's much easier to test works. 

Like how is "psbt-not-finalized" ever going to be useful to a sender?

Reply:
@RHavar yeah I think you are right. I will change that. In BTCPay SErver implementation I want to do what I can to help sender implementers to debug their stuff without exposing the receiver too much.

Reply:
Mmmh so it means I need to add a chapter explaining what the original PSBT should exactly contain.

Reply:
@RHavar can you review?

Reply:
Looks good 👍